plugins/view/src/types.ts

```diff
@@ -0,0 +1,776 @@
+  viewOptions?: ViewOptionsModel
+  variant?: string
+}
+
+/**
+ * @public
+ */
+export interface ViewletConfigOptions {
+  hiddenKeys?: string[]
+  strict?: boolean
+  extraProps?: Omit<BuildModelKey, 'key'>
+  sortable?: boolean
+}
+
+/**
+ * @public
+ */
+export interface LinkPresenter extends Doc {
+  pattern: string
+  component: AnyComponent
+}
+
+/**
+ * @public
+ *
+ *  "Alt + K" =\> Alt and K should be pressed together
+ *  "J T" - J and then T shold be pressed.
+ */
+export type KeyBinding = string
+
+/**
+ * @public
+ */
+export type ViewActionInput = 'focus' | 'selection' | 'any' | 'none'
+
+/**
+ * @public
+ */
+
+export type ViewActionFunction<T extends Doc = Doc, P = Record<string, any>> = (
+  doc: T | T[] | undefined,
+  evt?: Event,
+  params?: P
+) => Promise<void>
+
+/**
+ * @public
+ */
+export type ViewActionAvailabilityFunction<T extends Doc = Doc> = (doc: T | T[] | undefined) => Promise<boolean>
+
+/**
+ * @public
+ */
+export type ViewAction<P = Record<string, any>> = Resource<ViewActionFunction<Doc, P>>
+
+/**
+ * @public
+ */
+export interface ActionCategory extends Doc, UXObject {
+  // Does category is visible for use in popup.
+  visible: boolean
+}
+
+/**
+ * @public
+ */
+export type ActionGroup = 'create' | 'edit' | 'associate' | 'copy' | 'tools' | 'other' | 'remove'
+
+/**
+ * @public
+ */
+export interface Action<T extends Doc = Doc, P = Record<string, any>> extends Doc, UXObject {
+  // Action implementation details
+  action: Resource<ViewActionFunction<T, P>>
+  // Action implementation parameters
+  actionProps?: P
+
+  // If specified, will show sub menu based on actionPopup/actionProps
+  actionPopup?: AnyComponent
+
+  // If specified, action could be used only with one item selected.
+  // single - one object is required
+  // any - one or multiple objects are required
+  // any - any input is suitable.
+  input: ViewActionInput
+  inline?: boolean
+
+  // Focus and/or all selection document should match target class.
+  target: Ref<Class<Doc>>
+  // Action is applicable only for objects matching criteria
+  query?: DocumentQuery<T>
+
+  // Action is shown only if the check is passed
+  visibilityTester?: Resource<ViewActionAvailabilityFunction<T>>
+
+  // If defined, types should be matched to proposed list
+  inputProps?: Record<string, Ref<Class<Doc>>>
+
+  // Kayboard bindings
+  keyBinding?: KeyBinding[]
+
+  // short description for action.
+  description?: IntlString
+
+  // Action category, for UI.
+  category: Ref<ActionCategory>
+
+  // Context action is defined for
+  context: ViewContext
+
+  // A list of actions replaced by this one.
+  // For example, it could be global action and action for focus class, second one fill override first one.
+  override?: Ref<Action>[]
+
+  // Avaible only for workspace owners
+  secured?: boolean
+  allowedForEditableContent?: boolean
+}
+
+/** @public */
+export interface ActionWithAvailability<T extends Doc, P = never> {
```

Comment:
The only added type is this one. All other code was just moved from `index.ts`

Reply:
Why we need it at all? (I mean `ActionWithAvailbility`, because we already have visibilityTester in `Action`

Reply:
Basically, it's a sugar. It couples together visibility check and action handler, so that you don't have to copy-paste the check in handler, and can keep tightly related functions in the same place. And utility functions eliminate the need of boilerplate to handle too wide input types (`Doc` vs `Doc | Doc[] | undefined`). You can check out the real app use-cases in [`questions`](https://github.com/hcengineering/uberflow/tree/main/plugins/questions-resources/src/actions) or [`training`](https://github.com/hcengineering/uberflow/tree/main/plugins/training-resources/src/actions) plugins

Reply:
OK, removed `ActionWithAvailability` for now